core/src/main/java/org/testcontainers/containers/BaseContainerDef.java

```diff
@@ -0,0 +1,332 @@
+package org.testcontainers.containers;
+
+import com.github.dockerjava.api.command.CreateContainerCmd;
+import com.github.dockerjava.api.model.Bind;
+import com.github.dockerjava.api.model.ExposedPort;
+import com.github.dockerjava.api.model.HostConfig;
+import com.github.dockerjava.api.model.InternetProtocol;
+import com.github.dockerjava.api.model.PortBinding;
+import com.github.dockerjava.api.model.Ports;
+import lombok.Getter;
+import lombok.experimental.Delegate;
+import lombok.extern.slf4j.Slf4j;
+import org.testcontainers.DockerClientFactory;
+import org.testcontainers.UnstableAPI;
+import org.testcontainers.containers.wait.strategy.WaitStrategy;
+import org.testcontainers.images.RemoteDockerImage;
+import org.testcontainers.utility.DockerImageName;
+import org.testcontainers.utility.TestcontainersConfiguration;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@UnstableAPI
+@Slf4j
+abstract class BaseContainerDef {
```

Comment:
I started with `AbstractContainerDef` bud didn't feel right :D

Reply:
Why didn't it feel right? Not that I have a strong opinion either base. Bloch says in Effective Java, that the `Abstract` prefix is nowadays considered a convention from his PoV. I this the `Base` prefix is more conventional in the .NET world.

Reply:
I missed to put some context, there is nothing wrong about the name `AbstractContainerDef` but using it as a field in `GenericContainer`.